### PR TITLE
Actions: Add workflow for marking stale questions

### PR DESCRIPTION
**Pull request by aeisenberg**
This PR adds a workflow for marking and closing issues as stale. Issues must be labeled as _question_. PRs are never marked as stale.

### CODE REVIEW

File: .github/workflows/close-stale.yml
Comment (adityasharad) [inline .github/workflows/close-stale.yml:1]:
```suggestion
name: Mark stale issues
```

File: .github/workflows/close-stale.yml
Comment (adityasharad) [inline .github/workflows/close-stale.yml:21]:
```suggestion
        close-issue-message: 'This issue was closed because it has been inactive for 7 days.'
```

**Comment by aeisenberg** [comment]
I've set the `debug-only` flag to true, but looks like debug info is not being displayed. I think we need to set `ACTIONS_STEP_DEBUG` as a repo secret. Can someone with access set this secret and re-run the workflow?

File: .github/workflows/close-stale.yml
Comment (adityasharad) [inline .github/workflows/close-stale.yml:7]: I think we can drop the `pull_request` trigger and just do testing via `workflow_dispatch` on branches.
```suggestion
```

File: .github/workflows/close-stale.yml
Comment (adityasharad) [inline .github/workflows/close-stale.yml:12]: You may want to add `if: ${{ github.repository == 'github/codeql }}`  so that this doesn't run on forks.

**Review by adityasharad — COMMENTED** [review]
Looks reasonable, a few suggestions.

**Comment by adityasharad** [comment]
Added `ACTIONS_STEP_DEBUG=true`. Go ahead and rerun the workflow.

File: .github/workflows/close-stale.yml
Comment (aeisenberg) [reply .github/workflows/close-stale.yml]: path is incorrect also. I need to have this run once before it appears as an option for `workflow_dispatch`. I'll remove it later.

File: .github/workflows/close-stale.yml
Comment (aeisenberg) [reply .github/workflows/close-stale.yml:12]: Done

**Comment by aeisenberg** [comment]
Currently, the workflow will mark 120 issues as stale. I spot checked a few of them and they all do seem legitimately stale.

**Comment by aeisenberg** [comment]
I think this looks good. I see no problem running against all historic issues. Here are the issues that would have been closed. I'll make sure no other committers have an issue with this, before merging.

* #5501 
* #5480 
* #5472 
* #5451 
* #5441 
* #5405 
* #5404 
* #5403 
* #5399 
* #5392 
* #5376 
* #5343 
* #5331 
* #5320 
* #5314 
* #5298 
* #5297 
* #5291 
* #5290 
* #5289 
* #5288 
* #5274 
* #5213 
* #5177 
* #5115 
* #5104 
* #5084 
* #5079 
* #5054 
* #4988 
* #4937 
* #4930 
* #4910 
* #4909 
* #4902 
* #4890 
* #4883 
* #4870 
* #4852 
* #4851 
* #4843 
* #4840 
* #4815 
* #4804 
* #4788 
* #4699 
* #4651 
* #4591 
* #4551 
* #4530 
* #4496 
* #4470 
* #4426 
* #4379 
* #4365 
* #4284 
* #4280 
* #4180 
* #4168 
* #4145 
* #4128 
* #4119 
* #4034 
* #4005 
* #3967 
* #3963 
* #3949 
* #3908 
* #3903 
* #3827 
* #3826 
* #3825 
* #3818 
* #3799 
* #3711 
* #3710 
* #3709 
* #3707 
* #3705 
* #3698 
* #3696 
* #3695 
* #3649 
* #3648 
* #3647 
* #3644 
* #3642 
* #3635 
* #3624 
* #3623 
* #3605 
* #3493 
* #3461 
* #3428 
* #3418 
* #3417 
* #3406 
* #3292 
* #3290 
* #3289 
* #3266 
* #3261 
* #3252 
* #3102 
* #3101 
* #3100 
* #3017 
* #2954 
* #2952 
* #2826 
* #2775 
* #2774 
* #2749 
* #2714 
* #2548 
* #2508 
* #2477 
* #2424 
* #2337

**Comment by geoffw0** [comment]
Just scanned through the list above, looking mostly for C++ issues.  They're certainly stale, most are questions, but there are a significant number of mislabeled suggestions (perhaps the correct label for these is not obvious?).  Having something remind us to deal with these seems like a good idea.  I would want a notification before any automatic closure (I get the impression the PR does this).

What's the minimum I'll need to do to prevent something from being closed?  Will assigning it or changing labels work, or do I need to comment?

---

I've just dealt with:
- https://github.com/github/codeql/issues/4128 "CodeQL can't find the C++ API xmlFree from libxml2"
	- two follow up questions were asked but not answered.  Closed.
- https://github.com/github/codeql/issues/3292 "eed help for write query [cpp]"
	- I replied but the thread stopped there.  Bit surprising considering how much effort the author had put into writing it up, but I'm guessing the problem was solved or forgotten long ago.  Closed.
- https://github.com/github/codeql/issues/3290 "C++: Range analysis not picking up on condition because of implicit casts"
	- this was answered I think to the satisfaction of all.  Closed.
- https://github.com/github/codeql/issues/4651 "can codeQL analyze Broken Access Control vulnerabilities or unauthorized vulnerabilities?"
	- I asked some questions but was not answered.  We still don't know what the question was about.  Closed.
- https://github.com/github/codeql/issues/5274 "Some .qhelp files appear to be unused"
	- a suggestion not a question; added to my backlog

---

These look worthy of our attention, any volunteers?
- https://github.com/github/codeql/issues/4379 "Broken qhelp links"
- https://github.com/github/codeql/issues/5298 "Improve string predicates documentation"

**Comment by aeisenberg** [comment]
Thanks for going through the open issues. Now that we have the first responder rotation, I expect that we won't have any more of these mislabeled issues.

> I would want a notification before any automatic closure (I get the impression the PR does this).

Yes. There are two steps. The first is marking the issue as _stale_, which adds the label as well as a comment. The second step is to close the issue 7 days later.

> What's the minimum I'll need to do to prevent something from being closed? Will assigning it or changing labels work, or do I need to comment?

Either comment or remove the the stale label.

**Comment by geoffw0** [comment]
Sounds great. :+1:

**Review by adityasharad — DISMISSED** [review]

**Review by adityasharad — APPROVED** [review]

**Comment by Marcono1234** [comment]
What was the intention of adding this workflow? Was it in general the too high volume of issues, or was it that there were too many issues on which the reporter never responded back?
If the intention was to close issues without reporter response, then this workflow is in my opinion too destructive. A similar situation occurred in https://github.com/actions/stale and https://github.com/actions/starter-workflows, though they let it run for pull requests, and to be fair apparently had / have a bigger spam problem than this repository. In the end the bot closed a bunch of legit issues, where the maintainers never even responded in the first place, or where the reporter was waiting for the response of a maintainer.
And it looks like the same is happening here now, of the 15 issues currently [labeled as `stale`](https://github.com/github/codeql/issues?q=is%3Aissue+label%3Astale), #5441 seems to be the only one where you are actually waiting for the response of the reporter. In all other cases there has either been no interaction at all from you maintainers, or you told them that the topic will be discussed internally. I am not trying to blame you for not interacting with the issues, especially since I may be one of the major creators of issues, though on the other hand I am not very keen on having to comment every 14 - 21 days on my issues just to keep them open. And also there are good issues / feature requests (for which I am not the author) which would be closed as well.

If your intention is really only to resolve issues without reponse from the reporter, maybe it would be better to introduce an `awaiting-response` label or similar which you add after expecting a response from the reporter, and then only have the bot run for issues with this label; possibly also with a bot removing the label automatically if there has been any non-bot activity.

**Comment by aeisenberg** [comment]
Thanks for asking about this change. We've had a first responder rotation for the past couple of months. Each week one developer on our team becomes the point-person for answering questions (or forwarding to the proper team). The goal with this change is to make it clear what issues are currently unanswered or awaiting response.

Though, going through some more of the `stale` issues now, I see that some of them are incorrectly labeled as `questions`. This stale process only runs for questions (not bugs or enhancements). Part of the first-responder's job is to properly triage these new issues. Now that we have this automated process we need to be more precise about questions (relabeling and forwarding as appropriate).

Perhaps, as you suggest, we should be adding an `awaiting-response` label and running the stale process off of that.

**Comment by aschackmull** [comment]
@aeisenberg There are a number of issues (a bunch of which was opened by @Marcono1234) that I'd like to keep around without having to interact with them every 14-21 days just to keep them open. A bunch of these are backlogged in the Java team and we haven't yet had the capacity to triage them, so we can't really say whether they are best labelled as bugs or enhancements. How do you suggest we best keep them around? Simply remove the question label? (Leaving them perhaps with no labels at all).

**Comment by leonard84** [comment]
If you are using such an workflow, then you should improve the available templates, e.g., introduce a bug report template and so on. Because right now, everything that is not a false positive or vulnerability gets labeled as question, even if it is a bug report. 

> Perhaps, as you suggest, we should be adding an `awaiting-response` label and running the stale process off of that.

Sound like a good idea.

**Comment by aeisenberg** [comment]
See here #5689

**Pull request merged**
